body/stretch_body/stepper.py

```diff
@@ -656,6 +663,9 @@ def unpack_board_info(self,s):
         self.board_info['firmware_version'] = unpack_string_t(s[sidx:], 20).strip('\x00')
         self.board_info['protocol_version'] = self.board_info['firmware_version'][self.board_info['firmware_version'].rfind('p'):]
         sidx += 20
+
+        if int(self.board_info['protocol_version'].strip('p')) >= 5:
```

Comment:
Only the `unpack_board_info()` method isn't respecting the class override order.

Reply:
Hi @vrathiraj-hellorobot, I remember now why this happens. The reason is because `Stepper` does not know what protocol version the firmware will want to communicate with until after the board info has been unpacked. So overriding `unpack_board_info()` in `Stepper_Protocol_Px` subclasses is no good since that would require the class to already know which protocol to communicate with. It's a chicken and egg.

In more detail: When the user initializes a `Stepper` class (or more likely, they would initialize `Robot` which initializes `Stepper` internally), they are initializing a class that has only extended `StepperBase`. 

https://github.com/hello-robot/stretch_body/blob/250f2fbc461ee34361d9391d47d7d07e27e03869/body/stretch_body/stepper.py#L1375

Then, they call `startup()` which calls `StepperBase.startup()`.

https://github.com/hello-robot/stretch_body/blob/250f2fbc461ee34361d9391d47d7d07e27e03869/body/stretch_body/stepper.py#L1394-L1399 

Which initiates a "GET STEPPER BOARD INFO" RPC.

https://github.com/hello-robot/stretch_body/blob/250f2fbc461ee34361d9391d47d7d07e27e03869/body/stretch_body/stepper.py#L168-L175

`rpc_board_info_reply()` receives the firmware's reply and passes it along.

https://github.com/hello-robot/stretch_body/blob/250f2fbc461ee34361d9391d47d7d07e27e03869/body/stretch_body/stepper.py#L784-L786

Finally, `StepperBase.unpack_board_info()` gets the raw data and interprets it into the stepper_type, board_variant, hardware_id, firmware_version, and most importantly, protocol_version.

https://github.com/hello-robot/stretch_body/blob/250f2fbc461ee34361d9391d47d7d07e27e03869/body/stretch_body/stepper.py#L656-L669

Now the chain of methods return until we get back to `Stepper.startup()`. Importantly, `Stepper` now knows the protocol version the firmware wants to communicate with, so it overrides its own methods with the ones from the `Stepper_Protocol_Px` classes in ascending order.

https://github.com/hello-robot/stretch_body/blob/250f2fbc461ee34361d9391d47d7d07e27e03869/body/stretch_body/stepper.py#L1394-L1404  

This is why overriding `unpack_board_info()` in a `Stepper_Protocol_Px` class has no effect when the `Stepper` class is initialized.